### PR TITLE
feat: add Postgres libcluster strategy

### PR DESCRIPTION
**Pull request by w3b6x9**
## What kind of change does this PR introduce?

feature

## What is the current behavior?

Current cluster strategy is DNS polling.

## What is the new behavior?

New cluster strategy is Postgres LISTEN/NOTIFY (pub/sub) via heartbeat mechanism.

### CODE REVIEW

**Comment by chasers** [comment]
Another thought:

What happens if a node doesn't get a heartbeat in time, adds it to the bad_nodes or inactive_nodes list and then it starts heartbeating again? In the event where a node can't talk to Postgres but can talk to the other nodes? I guess at that point it would reconnect and send the sync again.

File: lib/realtime/cluster_strategy/postgres.ex [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+          new_inactive_nodes
+      end
+
+    new_state = %{
+      state
+      | meta: %{
+          meta
+          | inactive_nodes: new_inactive_nodes
+        }
+    }
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply lib/realtime/cluster_strategy/postgres.ex:224]: @abc3 tried this but doesn't work

File: lib/realtime/cluster_strategy/postgres.ex [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+
+    message =
+      case notify_event do
+        :sync -> "sync::#{node()}"
+        :heartbeat -> "heartbeat::#{node()}"
+      end
+
+    timeout =
+      Repo
+      |> SQL.query("NOTIFY #{@channel}, '#{message}'", [])
```
[the remark this answers is not in the document]
Comment (w3b6x9) [reply lib/realtime/cluster_strategy/postgres.ex:151]: @abc3 that's not possible b/c that conn is created in `Postgrex.Notifications` which creates its own `Postgrex.SimpleConnection` so there's no way to handle the `NOTIFY` query out of the box.

Let's try this out and if need be I'll use Postgrex or `Postgrex.SimpleConnection` (direct, no pooler) to create a separate connection for the `NOTIFY` query. I'd rather tap into the existing pool b/c we've assigned it a lot of connections.

**Comment by w3b6x9** [comment]
@abc3 can you take another look at this? i updated it according to your feedback and consolidated it into one commit.

**Comment by w3b6x9** [comment]
@chasers 

> Can you outline the intended logic for me to make sure I'm reading this right?

The gist of this is that it's kinda similar to how we do heartbeats and timeouts in Phoenix Channels. A node is going to subscribe to Postgres PubSub via PG LISTEN, and then every 4-6 seconds (random jitter), this can be customized, it'll reach out to Postgres via NOTIFY to send heartbeats out to all the other nodes that are also subscribed to via PG LISTEN.

Default timeout is 15 seconds so if Node A doesn't hear from Node B in that time then it will seek to disconnect from Node B. 

> Why are you using :gen_statem if the state is always :no_state?

It was the most elegant way of doing this without having to deal with multiple `handle_continue`s and a bunch of refs and canceling refs and `Process.send_after`s. `:gen_statem` handles that for you depending on the different events and has built in generic timeouts.

It's also my first `:gen_statem` and thought it was more straightforward if I went with `:no_state`. It seems similar to `Postgrex.SimpleConnection` which was refactored to use `:gen_statem` and it's also using `:no_state`.

> Why is the heartbeat timeout different on a notify success or error?

My intent was to be as aggressive as possible when things errored so wanted to retry as quickly as possible.

> Retrying between 4 - 6 seconds won't make any difference. Should really be an exponential backoff, but we also don't have a lot of procs here which will be hitting Postgres. It'll be N nodes doing this so it should hammer the db either way.

It's not retrying every 4-6 seconds but that's how often the heartbeats are sent to PG which are then sent to all the other nodes that have elected to PG LISTEN. This is like the Realtime clients sending heartbeats every 30 seconds to Realtime servers but wanted to introduce a bit of jitter since it's N nodes hammering the DB.

It could definitely be exponential backoff but wanted to keep it simple so just did random jitter. Didn't wanted add more state to the `MetaState`.

> It looks like you're using meta here as we'd normally use a genserver state. Can we have a struct for this to make it clear on what is expected there?

Sounds good, I added `MetaState`. Since we're dealing with `libcluster` State, the best place to put our custom state was in the meta slice of that State.

> Can we use binary pattern matching on the sync and heartbeat messages for control flow?

I attempted to do this after you mentioned it but after implementing @abc3 requested changes it's better for now to keep it as is. Open to refactoring that later I think it's alright for now.

> Can you explain the difference between inactive_nodes and bad_nodes?

There is no difference. The inactive_nodes are just used to track the no longer heart beating nodes so to cleanly remove them at a later point if it failed to do so initially.

> It seems that if a node looses connectivity it would not heartbeat which when would cause disconnect to error which would then put it in the new_inactive_nodes list and it would try to disconnect from that bad node now in a loop.

That disconnect is a result of the call to `:erlang.disconnect_node/1` and unrelated to PG NOTIFY/LISTEN. So if it failed then it would be tried again later. This is similar to what `libcluster` DNS poll strategy is doing.

> Is it just trying to connect to itself? The connect fn defined there always returns true so it thinks it's connected?

So it pretends that another node is sending sync/heartbeat and then checking to see if it received the `{:connect, other_node}` message in its mailbox. The connect fn does always return true but I'm checking the received message itself.

> What happens if a node doesn't get a heartbeat in time, adds it to the bad_nodes or inactive_nodes list and then it starts heartbeating again? In the event where a node can't talk to Postgres but can talk to the other nodes? I guess at that point it would reconnect and send the sync again.

Yea, so PG is the source of truth so when a node doesn't get a series of heartbeats from other node, then the node will flag it as inactive and seek to disconnect. If it does heartbeat it gets removed from inactive nodes list so it's not part of the disconnect nodes call.

File: lib/realtime/cluster_strategy/postgres.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (abc3) [reply lib/realtime/cluster_strategy/postgres.ex]: Hmm, never mind then

```elixir
iex(1)> state = %{some_key: 1, meta: %{inactive_nodes: [:a]}}
%{meta: %{inactive_nodes: [:a]}, some_key: 1}
iex(2)> put_in(state[:meta][:inactive_nodes], [:b])              
%{meta: %{inactive_nodes: [:b]}, some_key: 1}
```

File: config/runtime.exs [hunk 3]
```diff
@@ -111,6 +100,73 @@ if config_env() != :test do
...
 
+default_cluster_strategy =
+  config_env()
+  |> case do
+    :prod -> "DNS"
+    _ -> "EPMD"
+  end
+
+cluster_topologies =
+  System.get_env("CLUSTER_STRATEGIES", default_cluster_strategy)
```
Comment (abc3) [inline config/runtime.exs:111]: 🔥 🔥 🔥

**Review by abc3 — APPROVED** [review]
Looks good to me!

File: lib/realtime/cluster_strategy/postgres.ex (hunk 1 above)
[the remark this answers is not in the document]
Comment (w3b6x9) [reply lib/realtime/cluster_strategy/postgres.ex]: It’s bc it’s not a map but a struct. Libcluster state is %State{} and meta is %MetaState{} and put_in doesn’t work for structs.

**Comment by chasers** [comment]
@w3b6x9 sounds good. Can you just add the intended logic to the module doc please?

**Comment by w3b6x9** [comment]
> @w3b6x9 sounds good. Can you just add the intended logic to the module doc please?

@chasers added it: https://github.com/supabase/realtime/pull/582/files#diff-010b78e1c813da47ff100f0672f1ba1a87ee8b9c38cca49bfa28c44d15e20711R5-R13

**Pull request merged**